File: backend/app/schemas/reportes.py
from datetime import date

from pydantic import BaseModel


class EstudianteDisponible(BaseModel):
    id_estudiante: int
    name_estudiante: str
    sec_name_estudiante: str
    grupo: str | None = None


class ConteoEstado(BaseModel):
    estado: str
    cantidad: int


class ItemNota(BaseModel):
    """Una evaluacion con la nota del estudiante y su aporte a la nota final."""

    id_evaluacion: int
    name_evaluacion: str
    porcentaje: float
    fecha: date | None
    valor: float | None
    contribucion: float | None  # valor * porcentaje / 100


class RubroReporte(BaseModel):
    tipo: str  # Examen | Tarea | Cotidiano
    peso: float  # suma de los % de sus evaluaciones
    items: list[ItemNota]
    contribucion: float | None  # suma de los aportes de las evaluaciones calificadas


class AsistenciaRubro(BaseModel):
    peso: float  # % de la asistencia en la nota (porcentaje_asistencia de la clase)
    total_registros: int
    presentes: int  # registros que NO son "Ausente"
    porcentaje_presente: float | None  # nota de asistencia (0-100)
    contribucion: float | None
    por_estado: list[ConteoEstado]


class PeriodoReporte(BaseModel):
    periodo: int
    rubros: list[RubroReporte]
    asistencia: AsistenciaRubro
    peso_total: float  # suma de todos los pesos (rubros + asistencia)
    nota_periodo: float | None  # nota ponderada (aportes calificados + asistencia)


class MateriaReporte(BaseModel):
    id_clase: int  # id_profesor_asignatura_grupo
    materia: str
    profesor: str
    porcentaje_asistencia: float
    periodos: list[PeriodoReporte]
    nota_final: float | None  # promedio de las notas de los periodos calificados


class ReporteEstudiante(BaseModel):
    id_estudiante: int
    name_estudiante: str
    sec_name_estudiante: str
    grupo: str | None
    materias: list[MateriaReporte]

File: tailwind-admin-reactjs-free/package/src/types/reportes.ts
export interface EstudianteDisponible {
  id_estudiante: number;
  name_estudiante: string;
  sec_name_estudiante: string;
  grupo: string | null;
}

export interface ConteoEstado {
  estado: string;
  cantidad: number;
}

export interface ItemNota {
  id_evaluacion: number;
  name_evaluacion: string;
  porcentaje: number;
  fecha: string | null;
  valor: number | null;
  contribucion: number | null;
}

export interface RubroReporte {
  tipo: string; // Examen | Tarea | Cotidiano
  peso: number;
  items: ItemNota[];
  contribucion: number | null;
}

export interface AsistenciaRubro {
  peso: number;
  total_registros: number;
  presentes: number;
  porcentaje_presente: number | null;
  contribucion: number | null;
  por_estado: ConteoEstado[];
}

export interface PeriodoReporte {
  periodo: number;
  rubros: RubroReporte[];
  asistencia: AsistenciaRubro;
  peso_total: number;
  nota_periodo: number | null;
}

export interface MateriaReporte {
  id_clase: number;
  materia: string;
  profesor: string;
  porcentaje_asistencia: number;
  periodos: PeriodoReporte[];
  nota_final: number | null;
}

export interface ReporteEstudiante {
  id_estudiante: number;
  name_estudiante: string;
  sec_name_estudiante: string;
  grupo: string | null;
  materias: MateriaReporte[];
}
